### PR TITLE
Fix modify image

### PR DESCRIPTION
**Pull request by buthanoid**
I noticed some bugs in the last issue modify-image  https://github.com/JMMC-OpenDev/oimaging/issues/16

1. When loading an `OifitsFile` for using it in the input form, we should not refuse `HDUs` from their name, because we know nothing about the file we are loading. We must use checksum. See also https://github.com/buthanoid/oimaging/blob/documentation/doc/ajout-image-oi-fits-image-hdu-library.md for more details.
2. The computation of the variable `checkSelectedImageHDUs` in function `addFitsImageHDUs` was erroneous : a null list would mark for checking whereas it should mark for no-checking.

I am still checking for more bugs, I will fix them here if any

### CODE REVIEW

**Comment by buthanoid** [comment]
Did not see any more bug. However I think another design could be simpler, and would not need doc anymore. https://github.com/buthanoid/oimaging/blob/documentation/doc/ajout-image-oi-fits-image-hdu-library.md -> Autre design

**Comment by bourgesl** [comment]
To answer one of questions, IR software will rewrite input images (INIT, RGL_PRIOR) differently (murphy's law) so the keyword list will be different (order, comments, different header) and then checksum(result.inputImages) != checksum (input.inputImages), that's why the HDU_NAME was used to check for uniqueness (but it is not good enough).

An alternative consist in building our own identity check on image HDU: important keywords as numeric values (area + array dims) and of course using datasum (over flux values) => identity sum (based on ordered keyword values in known unit and precision + data sum) could be better.

In conclusion: the FITS checksum is too strict as any change in FITS header (unit, comment, rounded values % numerical precision) implies the images are considered as different although these images are same in fact.

**Comment by FerreolS** [comment]
I think it is also a OI-interface specification problem. If the software modify an input image, it should warn somehow the user. Should I raise an issue there?

**Comment by buthanoid** [comment]
Should they be allowed to do so ? In what lengths ? Adding few pixels, mirroring the image... 

Anyway, if they are doing it right now, I suppose we must handle it.

I don't know how to warn the user. A message dialog sounds a bit too annoying to me, what do you think ? Maybe using the message info box ?

**Comment by FerreolS** [comment]
> Should they be allowed to do so ? In what lengths ? Adding few pixels, mirroring the image...

Nobody ever think about this case. I think it should be forbidden or indicated otherwise with a flag or something.
 
> Anyway, if they are doing it right now, I suppose we must handle it.
> 
> I don't know how to warn the user. A message dialog sounds a bit too annoying to me, what do you think ? Maybe using the message info box ?

As a warning, I was just think as a flag in the OI-interface format nothing in the GUI

> 
> If we have a "todo one day" list, we could add the topic "working on a new identity for FitsImageHDU"

**Comment by bourgesl** [comment]
> I think it is also a OI-interface specification problem. If the software modify an input image, it should warn somehow the user. Should I raise an issue there?
I don't think so, as it is a technical problem.

The issue consists in the fact that IR software are using different FITS / OIFITS libraries (idl astrolib, cfitsIO, julia ...) and typically read Fits image HDUs from the input file and write them again in the output file (copy): due to murphy's law, possible changes happen (in reality) like keyword list order modified, conversions (units), rounding of floating-point values ... and the image pixels can be converted into any type (int, float, double) so checksum(output.input image) != checksum(input.input image) whereas images are the 'same' (within numerical precision...)
It is a serialization issue (FITS) and reproducibility problem that is hard to handle properly as IR software are maintained by third-parties.

**Comment by bourgesl** [comment]
> Did not see any more bug. However I think another design could be simpler, and would not need doc anymore. https://github.com/buthanoid/oimaging/blob/documentation/doc/ajout-image-oi-fits-image-hdu-library.md -> Autre design

As the logic seems clear to you, please try this 'simpler' and more explicit approach ?

**Comment by buthanoid** [comment]
I found an erroneous scenario:
- add HDU "AAA" to the image library
- add HDU "BBB" to the image library
- select "AAA" as init image
- run WISARD
- before Run is finished, select "BBB" as init image
- Run finish and tries to refuse quickly its "AAA" init image. But it fails because init image of the input form is now "BBB" and not "AAA" anymore. So it renames "AAA" to "AAA-another-sad-wednesday-11:41" and add it to the library.

**Comment by bourgesl** [comment]
This puzzle is going to make me crazy.
Let's be more obvious for both dev & end-user: the image library contains all 'distinct' images (no duplicates in the sens ~ equivalent). I propose to have an explicit mappings in both directions:
- Map<FitsImageHDU, FitsImageHDU> ImageHDUtoImageLib : all FitsImageHDU loaded (from files) to the corresponding 'distinct' FitsImageHDU in the ImageLib
- Map<FitsImageHDU, List<FitsImageHDU>> ImageLibtoImageHDU : all 'distinct' FitsImageHDU in the ImageLib to their corresponding FitsImageHDUs loaded (from files) 
The second map is not really useful for now.

At any time and in particular while loading an OIFITS input file, you can make a lookup in the ImageHDUtoImageLib to get the FitsImageHDU in the ImageLib to be selected corresponding to the file's keywords (INIT_IMG or RGL_PRIOR or LAST_IMG)

Of course, these mappings could be helpful to indicate image relationships in tooltips or in a new image table ?

**Comment by bourgesl** [comment]
I already faced this mapping problem in oifits explorer to map targets, instrument modes ... using fuzzy matchers, see:
https://github.com/JMMC-OpenDev/oitools/blob/master/src/main/java/fr/jmmc/oitools/model/AbstractMapper.java

**Comment by FerreolS** [comment]
Should we prevent user to change hdu during processing?

**Comment by buthanoid** [comment]
I moved some checking and renaming to the function `addToImageLibrary`

Note that some parts remains outside it. For example, `prepareImage()` is meant to be done, no matter we add the image to the library or not.

**Comment by buthanoid** [comment]
Replaced the list of HDU NAMES with a list of HDUS (references).

Note that is does not work when the oifitsfile contains duplicate HDU_NAME (or it does not work as intended). I propose to clean the oifitsfile in IRModel.postProcess() so that we can assume that the spec is respected in the other functions.

I changed the list reference comparison to a boolean so it is explicit and won't break if we change the list reference for some reason

**Comment by buthanoid** [comment]
I added a computation of the roles of the HDUs based on the spec and our discussions. I hope it is consistent with these.

It is then optionally used in addFitsImageHDUs to early skip INIT and RGL hdus (if a boolean is true). I added the boolean because only the Run scenario uses the early skipping.

I also modified addImageToLibrary to make it return the equivalent HDU in the library.

**Comment by buthanoid** [comment]
BUG:
When we update viewport (via viewport button or modify image),
the keywords are not updated (it can be seen via button "display keywords"). However if we save the  image in a file and load it, the keywords become correct.

**Comment by buthanoid** [comment]
BUG 2:
When we resample the image, the cross is not centered anymore.

**Comment by buthanoid** [comment]
Setting aside the two previous BUGS remarks (I should open another issue for these) ;
What we need to do to close this pull request is:
- When loading an oifitsfile in order to fill the input form. In the form we can only put images present in the library. So, first we fill the form with images from the library that are equivalents to the images in the oifitsfile. Then, secondly, we need to sync the file with the input form. So I think we must replace the original HDUS in the file with the new equivalent ones from the library, and modify the HDU_NAMEs in the keywords accordingly
- I am looking if there is something else

**Comment by buthanoid** [comment]
- I separated the selection from the function `addFitsImageHDUs`. 
- I made the function `addFitsImageHDUs` return the list of library equivalent hdus.
- In Run scenario I manually selected Result equivalent HDU as Initial image, and Rgl equivalent hdu as rgl image.
- In Load Oifitsfile scenario I manually selected Init equivalent HDU as initial image, and Rgl equivalent hdu as rgl image.
- In Load images scenario I manually selected Init equivalent HDU as initial image,
- In Modify images scenario I kept the code which replaces INIT or RGL if they were the origin of the modified image.
- For the previous selections, I used the roles to know which image should be selected
- I kept the call `updateImageIdentifiers` between `addFitsImageHDU` so it gets the correct names, and before `setSelectedInputImageHDU` so it applies to all HDUs (before the cleaning). It is a fragile point.

**Comment by bourgesl** [comment]
Looks good, 2 bugs fixed at once !
